localization/pose_instability_detector/src/pose_instability_detector.cpp

```diff
@@ -166,13 +170,267 @@ void PoseInstabilityDetector::callback_timer()
+
+  const double dead_reckoning_vertical_process_noise = dead_reckoning_lateral_process_noise;
+
+  const double dead_reckoning_angular_variance =
+    (twist_buffer_size - 1) * angular_velocity_variance_ * interval_sec * interval_sec;
+  const double dead_reckoning_angular_process_noise = 3 * sqrt(dead_reckoning_angular_variance);
+
+  // Set thresholds
+  threshold_diff_position_x_ = longitudinal_difference + dead_reckoning_longitudinal_process_noise +
+                               pose_estimator_longitudinal_tolerance_;
+  threshold_diff_position_y_ =
+    lateral_difference + dead_reckoning_lateral_process_noise + pose_estimator_lateral_tolerance_;
+  threshold_diff_position_z_ = vertical_difference + dead_reckoning_vertical_process_noise +
+                               pose_estimator_vertical_tolerance_;
+  threshold_diff_angle_x_ =
+    roll_difference + dead_reckoning_angular_process_noise + pose_estimator_angular_tolerance_;
+  threshold_diff_angle_y_ =
+    pitch_difference + dead_reckoning_angular_process_noise + pose_estimator_angular_tolerance_;
+  threshold_diff_angle_z_ =
+    yaw_difference + dead_reckoning_angular_process_noise + pose_estimator_angular_tolerance_;
+}
+
+void PoseInstabilityDetector::dead_reckon(
+  PoseStamped::SharedPtr & initial_pose, const rclcpp::Time & end_time,
+  const std::deque<TwistWithCovarianceStamped> & twist_deque, Pose::SharedPtr & estimated_pose)
+{
+  // get start time
+  rclcpp::Time start_time = rclcpp::Time(initial_pose->header.stamp);
+
+  // initialize estimated_pose
+  estimated_pose->position = initial_pose->pose.position;
+  estimated_pose->orientation = initial_pose->pose.orientation;
+
+  // define lambda function to convert quaternion to rpy
+  auto quat_to_rpy = [](const Quaternion & quat) {
+    tf2::Quaternion tf2_quat(quat.x, quat.y, quat.z, quat.w);
+    tf2::Matrix3x3 mat(tf2_quat);
+    double roll{};
+    double pitch{};
+    double yaw{};
+    mat.getRPY(roll, pitch, yaw);
+    return std::make_tuple(roll, pitch, yaw);
+  };
+
+  // cut out necessary twist data
+  std::deque<TwistWithCovarianceStamped> sliced_twist_deque =
+    clip_out_necessary_twist(twist_deque, start_time, end_time);
+
+  // dead reckoning
+  rclcpp::Time prev_odometry_time = rclcpp::Time(sliced_twist_deque.front().header.stamp);
+
+  for (size_t i = 1; i < sliced_twist_deque.size(); ++i) {
+    const rclcpp::Time curr_time = rclcpp::Time(sliced_twist_deque[i].header.stamp);
+    const double time_diff_sec = (curr_time - prev_odometry_time).seconds();
+
+    const Twist twist = sliced_twist_deque[i - 1].twist.twist;
+
+    // quat to rpy
+    auto [ang_x, ang_y, ang_z] = quat_to_rpy(estimated_pose->orientation);
+
+    // average quaternion of two frames
+    tf2::Quaternion average_quat;
+    average_quat.setRPY(
+      ang_x + 0.5 * twist.angular.x * time_diff_sec, ang_y + 0.5 * twist.angular.y * time_diff_sec,
+      ang_z + 0.5 * twist.angular.z * time_diff_sec);
```

Comment:
We should not convert quaternions to Euler angles, as it could lead to incorrect integral results when roll and pitch are non-zero.